### PR TITLE
Refactor the iteration code

### PR DESCRIPTION
**Pull request by WaDelma**
The parallel iterator PR that was merged was originally supposed to just contain documentation changes and this refactor that I wanted to do. I got bit sidetracked when I realised that I could make the splitting better.

(I just love how closures can be used to do these kinds of refactors*)

*and then you move the closure to a method

### CODE REVIEW

File: src/util.rs [hunk 1]
```diff
@@ -56,6 +57,17 @@ pub fn offsets(bit: Index) -> (usize, usize, usize) {
     (bit.offset(SHIFT1), bit.offset(SHIFT2), bit.offset(SHIFT3))
 }
 
+/// Gets the `usize` corresponding to layer and index
+pub fn get_from_layer<T: BitSetLike>(set: &T, layer: usize, idx: usize) -> usize {
```
Comment (torkleyy) [inline src/util.rs]: Could add this to `BitSetLike` with a default impl

**Review by torkleyy — COMMENTED** [review]
Hah! Nice ;)

File: src/iter/mod.rs [hunk 2]
```diff
@@ -38,61 +38,45 @@ impl<T: BitSetLike> BitIter<T> {
...
-                self.masks[2] &= !(1 << bit);
-                // Calculate index of the bit in second level
-                let idx = self.prefix[2] | bit;
-                // Take corresponding usize from layer below
-                self.masks[1] = self.set.layer1(idx as usize);
-                // Prefix of the index of the second level
-                self.prefix[1] = idx << BITS;
-                continue;
+        use self::State::*;
+        let mut handle_level = |level: usize| if self.masks[level] == 0 {
```
Comment (torkleyy) [inline src/iter/mod.rs]: I thought that I made this comment before already, but somehow it didn't work. Please make this a method on the struct; such a long closure makes the source code harder to understand and read.

**Review by torkleyy — APPROVED** [review]
This is great, thank you! I saw the code duplication before and this is exactly what I had in mind :)

**Review by kvark — APPROVED** [review]
neat!
bors r+

**Pull request merged**